src/BuildingBlocks/EventBus/InMemoryEventBus.cs

```diff
@@ -13,46 +12,50 @@ static InMemoryEventBus()
 
         private InMemoryEventBus()
         {
-            _handlers = new List<HandlerSubscription>();
+            _handlersDictionary = new Dictionary<string, List<IIntegrationEventHandler>>();
         }
 
         public static InMemoryEventBus Instance { get; } = new InMemoryEventBus();
 
-        private readonly List<HandlerSubscription> _handlers;
+        private readonly IDictionary<string, List<IIntegrationEventHandler>> _handlersDictionary;
 
         public void Subscribe<T>(IIntegrationEventHandler<T> handler)
             where T : IntegrationEvent
         {
-            _handlers.Add(new HandlerSubscription(handler, typeof(T).FullName));
+            var eventType = typeof(T).FullName;
+            if (eventType != null)
+            {
+                if (_handlersDictionary.ContainsKey(eventType))
+                {
+                    var handlers = _handlersDictionary[eventType];
+                    handlers.Add(handler);
+                }
+                else
+                {
+                    _handlersDictionary.Add(eventType, new List<IIntegrationEventHandler> { handler });
+                }
+            }
         }
 
         public async Task Publish<T>(T @event)
             where T : IntegrationEvent
         {
-            var eventType = @event.GetType();
+            var eventType = @event.GetType().FullName;
+
+            if (eventType == null)
```

Comment:
when the condition is true? I think only when the argument is null so we should add null check to `@event?.GetType().FullName`.

Reply:
`FullName` can be `null` in some cases: https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname?view=net-5.0#System_Type_FullName

Of course in our case shouldn't be `null`, but still, it is better to add this null check imo.